checker/src/main/java/org/checkerframework/checker/determinism/DeterminismVisitor.java

```diff
@@ -593,8 +594,54 @@ protected void checkConstructorInvocation(
         super.checkConstructorInvocation(invocation, constructor, newClassTree);
     }
 
+    /**
+     * Issue a warning if the under the following conditions:
```

Comment:
I don't understand why these rules were chosen. For example, why is the following constructor illegal?
```
@NonDet Object nonDetField;
@Det Object detField;
@PolyDet MyClass(@NonDet Object nonDet, @Det Object det) {
    nonDetField = nonDet;
    detField = det;
}
```

Reply:
@smillst According to these rules, this constructor will be illegal because if the return type is `@PolyDet`, then at least one of the parameters must be `@PolyDet`. May be the wording in the documentation isn't clear?

Reply:
I'm asking why it must have at least on @PolyDet parameter.   (The documentation is clear that the constructor is illegal, thought not the pull request title.)

Related is the following illegal?
```
@NonDet Object nonDetField;
@Det Object detField;
@Det MyClass(@NonDet Object nonDet, @Det Object det) {
    nonDetField = nonDet;
    detField = det;
}
```

Reply:
> I'm asking why it must have at least one `@PolyDet` parameter. (The documentation is clear that the constructor is illegal, thought not the pull request title.)

I see. If none of the parameters are polymorphic, then the return constructor type resolves to `@NonDet`. So you are right, I guess we can get rid of this check.

> Related, why is the following illegal?

This constructor should be illegal because of the non-deterministic assignment in it. For instance, let's say `x = new MyClass(@NonDet Object, @Det det);` Now if we print x, we will get non-deterministic results.

Reply:
> Now if we print x, we will get non-deterministic results.

How?  The return type of `toString` is `@PolyDet`, so `nonDetField` can't contribute to that String.

Reply:
```
@NonDet Object nonDetField;
@Det Object detField;
@Det MyClass(@NonDet Object nonDet, @Det Object det) {
    nonDetField = nonDet;
    detField = det;
}
void test1(@NonDet Object nd, @Det Object d) {
    MyClass x = new MyClass(nd, d);
    System.out.println(x);
}
```
Without the check, `x` in the above example will be `@Det`. Printing it will not result in any error. But, we know that the output will be non-deterministic because of the `@NonDet` field.
Having this check will ensure that we flag an error at the constructor which cannot be verified as `@Det`.

Reply:
`System.out.println(x)`  prints something like `MyClass@12313AD`, because MyClass doesn't override toString.  The Determinism Checker issues an error on that line. 
```
CheckConstructorReturn.java:18: error: [argument.type.incompatible] incompatible types in argument.
        System.out.println(new CheckConstructorReturn());
                           ^
  found   : @NonDet CheckConstructorReturn
  required: @Det Object
```

Reply:
> Your example does show that the checker needs to enforce that the return type of `toString` of objects passed to System.out.println is `@Det` (or `@PolyDet`).

The checker already enforces that `System.out.println` takes only `@Det` arguments (JDK is annotated according to this rule). The problem with not enforcing that a `@Det` constructor takes only `@Det` arguments is that we will wrongly assume that the argument to `System.out.println` (`x` in this example) is `@Det` when it is actually not.